src/ring_theory/unique_factorization_domain.lean

```diff
@@ -430,6 +448,218 @@ begin
+  { apply dvd_of_mk_le_mk, rw hp, exact dvd_of_mem_factors' (multiset.mem_inter.mp hp0).right }
+end
+
+theorem coprime_iff_inf_one {a b : α} (ha0 : a ≠ 0) (hb0 : b ≠ 0)
+  : (associates.mk a) ⊓ (associates.mk b) = 1 ↔ ∀ {d : α}, d ∣ a → d ∣ b → ¬ prime d :=
+begin
+  split,
+  { intros hg p ha hb hp,
+    have hp1 : irreducible p, { apply irreducible_iff_prime.mpr hp },
+    have hp2 : irreducible (associates.mk p), { apply (irreducible_mk_iff _).2 hp1 },
+    have ha' : (⟨associates.mk p, hp2⟩ : {a : associates α // irreducible a })
+      ∈ factors' a ha0, exact mem_factors'_of_dvd ha0 hp1 ha,
+    have hb' : (⟨associates.mk p, hp2⟩ : {a : associates α // irreducible a })
+      ∈ factors' b hb0, exact mem_factors'_of_dvd hb0 hp1 hb,
+    revert hg,
+    change ((factors (associates.mk a)) ⊓ (factors (associates.mk b))).prod ≠ 1,
+    rw ← multiset.prod_zero, apply ne.elim, apply mt (@eq_of_prod_eq_prod _ _ _inst_2 _ 0),
+    rw factors_mk _ ha0, rw factors_mk _ hb0,
+    rw ← with_top.coe_inf, apply (mt with_top.coe_eq_coe.mp),
+    apply mt multiset.eq_zero_iff_forall_not_mem.mp, apply not_forall.mpr,
+    use ⟨associates.mk p, hp2⟩, apply not_not.mpr,
+    rw multiset.inf_eq_inter,
+    apply multiset.mem_inter.mpr
+      (and.intro (mem_factors'_of_dvd ha0 hp1 ha) (mem_factors'_of_dvd hb0 hp1 hb)) },
+  { intro hc,
+    by_contradiction hg,
+    obtain ⟨p, ⟨hp, ⟨hpa, hpb⟩⟩⟩ := exists_prime_dvd_of_not_inf_one ha0 hb0 hg,
+    apply absurd hp (hc hpa hpb) }
+end
+
+theorem prime_pow_factors (p : associates α) (h₂ : irreducible p)
+ (k : ℕ) : factors (p ^ k) = some (multiset.repeat ⟨p, h₂⟩ k) :=
+begin
+  apply eq_of_prod_eq_prod, rw associates.factors_prod,
+  unfold factor_set.prod, rw multiset.map_repeat,
+  rw multiset.prod_repeat, refl
+end
+
+theorem dvd_prime_pow_iff_count_factors (m : associates α) (h₁ : m ≠ 0) (p : associates α)
+  (h₂ : irreducible p) (k : ℕ) : p ^ k ≤ m ↔ k ≤ count ⟨p, h₂⟩ m.factors :=
+begin
+  obtain ⟨a, nz, h⟩ := associates.exists_non_zero_rep h₁,
+  erw [← h, (factors_mk a nz), ← with_top.some_eq_coe, multiset.le_count_iff_repeat_le],
+  split,
+  { intro h3,
+    apply with_top.coe_le_coe.1,
+    rw [← with_top.some_eq_coe, ← (prime_pow_factors p h₂), ← (factors_mk a nz)],
+    exact factors_le.2 h3 },
+  { intro h3,
+    apply factors_le.1,
+    erw [factors_mk a nz, prime_pow_factors p h₂],
+    exact with_top.coe_le_coe.2 h3 }
+end
+
+theorem count_product {a : associates α} (ha : a ≠ 0) {b : associates α} (hb : b ≠ 0)
+  {p : { a : associates α // irreducible a }}
+  : count p a.factors + count p b.factors = count p (factors (a * b)) :=
+begin
+  obtain ⟨a0, nza, ha'⟩ := exists_non_zero_rep ha,
+  obtain ⟨b0, nzb, hb'⟩ := exists_non_zero_rep hb,
+  erw [factors_mul, ← ha', ← hb', factors_mk a0 nza, factors_mk b0 nzb],
+  rw [← factor_set.coe_add, ← with_top.some_eq_coe, ← with_top.some_eq_coe, ← with_top.some_eq_coe],
+  unfold count, rw multiset.count_add
+end
```

Comment:
Some changes:

 * `*` is called `mul`, so the theorem should be called `count_mul`
 * the lemma is more useful for `simp`/`rw` if the equality is reversed
 * some proof golfing

```suggestion
theorem count_mul {a : associates α} (ha : a ≠ 0) {b : associates α} (hb : b ≠ 0)
  {p : { a : associates α // irreducible a }} :
  count p (factors (a * b)) = count p a.factors + count p b.factors :=
begin
  obtain ⟨a0, nza, ha'⟩ := exists_non_zero_rep ha,
  obtain ⟨b0, nzb, hb'⟩ := exists_non_zero_rep hb,
  rw [factors_mul, ← ha', ← hb', factors_mk a0 nza, factors_mk b0 nzb, ← factor_set.coe_add,
      ← with_top.some_eq_coe, ← with_top.some_eq_coe, ← with_top.some_eq_coe, count_some,
      multiset.count_add],
  refl
end
```